[BD-38][TNL-8438] Replace settings button with link for non-supported apps with legacy links

kewne opened the pull request:
# Description

If an app has a legacy link and has no explicit support in the MFE,
a link is now rendered instead of the settings button.

After marking the `calculator` and `discussion` apps as supported, and running the backend with the changes in edx/edx-platform#28098, the page grid now looks like this:

![image](https://user-images.githubusercontent.com/3984777/126894448-c7bede82-7d23-451d-834a-669d1c4ddfe3.png)

Related tickets:
* [TNL-8438](https://openedx.atlassian.net/browse/TNL-8438)
* [BB-4416 (OpenCraft Internal)](https://tasks.opencraft.com/browse/BB-4416)

# Testing

Using the backend changes in edx/edx-platform#28098:

1. Open the pages and resources page for the demo course
2. Verify that:
  1. the settings button appears for "supported" apps (e.g. `calculator` and `discussion`)
  2. the link to studio appear for non-supported apps with legacy links (e.g. `texbooks`)
  3. Nothing appears for non-supported apps without legacy links (e.g. `teams`)

edx-webhook commented:
Thanks for the pull request, @kewne! I've created [BLENDED-916](https://openedx.atlassian.net/browse/BLENDED-916) to keep track of it in Jira.   More details are on the [BD-14](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1545011241/BD-14+Blockstore+Powered+Content+Libraries+Taxonomies) project page.


When this pull request is ready, tag your edX technical lead.


<!-- comment:welcome-blended -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

kewne commented on src/pages-and-resources/data/supportedPages.js:
```diff
@@ -0,0 +1,2 @@
+const supportedPages = new Set(['calculator', 'discussion']);
+export default supportedPages;
```
I feel this construction is very clunky, suggestions are very welcome.

xitij2000 replied:
I don't think this is needed at all. The page will either have a settings link, or a legacy link, or no link at all. If it has a legacy link, show that, if it doesn't show the settings link.

kewne replied:
Done!

xitij2000 commented on src/pages-and-resources/pages/PageCard.jsx:
```diff
@@ -33,6 +34,34 @@ function PageCard({
   };
 
+  const settingsButton = () => {
+    if (page.hasFrontendSupport) {
+      return (
+        <IconButton
+          className="mb-0 mr-1"
+          src={Settings}
+          iconAs={Icon}
+          size="inline"
+          alt={intl.formatMessage(messages.settings)}
+          onClick={() => history.push(`${pagesAndResourcesPath}/${page.id}/settings`)}
+        />
+      );
+    }
```
I don't think this is needed. If the page has a legacyLink, show that, if it doesn't show the settings button based on the existing logic.

xitij2000 reviewed (commented):
I think this mostly looks, good, but the scope is smaller than what you expected :-)

xitij2000 commented:
@kewne Please tag this with bd-38, not 14.

xitij2000 commented on src/pages-and-resources/pages/PageCard.jsx:
```diff
@@ -33,6 +33,31 @@ function PageCard({
+      );
+    }
+    return (
+      <IconButton
+        className="mb-0 mr-1"
+        src={Settings}
+        iconAs={Icon}
+        size="inline"
+        alt={intl.formatMessage(messages.settings)}
+        onClick={() => history.push(`${pagesAndResourcesPath}/${page.id}/settings`)}
+      />
+    );
```
Please use the forward arrow icon from here: https://edx.github.io/paragon/foundations/icons/

xitij2000 reviewed (approved):
:+1: This looks good. Just two suggestions. 
1. Make the entire card a link instead of just the icon
2. Use the correct icon.
Other than that it's good to go from my side. 
- [x] I tested this: tested on devstack with related PR. 
- [x] I read through the code
- [x] I checked for accessibility issues

kewne commented:
@xitij2000 
> Make the entire card a link instead of just the icon

Are you sure? This makes the UX a bit weird: if it's a legacy link, the whole card becomes a link but, if it's not, you have to click just the settings button.

xitij2000 commented:
> Are you sure? This makes the UX a bit weird: if it's a legacy link, the whole card becomes a link but, if it's not, you have to click just the settings button.

You're probably right, let's keep it the way you're implemented it.

mehaknasir commented on src/pages-and-resources/pages/PageCard.jsx:
```diff
@@ -33,6 +33,31 @@ function PageCard({
   };
 
+  const settingsButton = () => {
+    if (page.legacyLink) {
+      return (
+        <Hyperlink destination={page.legacyLink}>
+          <Icon
+            className="mb-0 mr-1"
+            src={ForwardArrow}
+            size="inline"
+            screenReaderText={intl.formatMessage(messages.settings)}
+          />
+        </Hyperlink>
+      );
+    }
+    return (
+      <IconButton
+        className="mb-0 mr-1"
+        src={Settings}
+        iconAs={Icon}
+        size="inline"
+        alt={intl.formatMessage(messages.settings)}
+        onClick={() => history.push(`${pagesAndResourcesPath}/${page.id}/settings`)}
+      />
+    );
```
what I suggest is to have a single return statement and make the logic inside the return like this 

`page.legacyLink ?  <Hyperlink destination={page.legacyLink}>  :  <IconButton>`

however what you have done is totally fine.

mehaknasir reviewed (approved):
just a suggestion added. otherwise looks good to me

awaisdar001 reviewed (changes requested):
The icon is slightly different than added in the PR. Can you please change it? 

![image (2)](https://user-images.githubusercontent.com/4252738/125963655-164f5514-f01d-49a5-a1f4-25b8931728e9.png)

mehaknasir commented:
can you please rebase this branch to resolve conflicts so I can run a workflow on it?

awaisdar001 commented:
> can you please rebase this branch to resolve conflicts so I can run a workflow on it?

I can run the testing workflow. 👍

awaisdar001 commented on src/pages-and-resources/pages/PageCard.jsx:
```diff
@@ -45,16 +70,7 @@ function PageCard({
-              onClick={() => history.push(`${pagesAndResourcesPath}/${page.id}/settings`)}
-            />
-          )}
+          && settingsButton()}
```
I disagree that we need to bound "external url direction" with `allowedOperations`. Currently the behavior is bound to show `->` Iff certain param in allowed operations are enabled. Which should not be the case for the app who uses external links. 

I would recommend to decouple that logic.

awaisdar001 commented:
Can you also please 
1. Rebased your branch?
2. Add updated screenshot?

kewne commented:
@awaisdar001 @mehaknasir Thank you for your comments; I think I've addressed all of your comments but let me know if I missed something.

Note: I've updated the screenshot in my first PR comment.
